File: matlab/me_arm.m
close all
clear
clc

L(1) = Link([0 7.5 2+8 pi/2 0]);
L(2) = Link([0 0 9 0 ]);
L(3) = Link([0 0 9 0 ]);
% L(4) = Link([0 0 7.5 0]);
me_arm = SerialLink(L,'name', 'me_arm');
me_arm

%range of motion        corresponding servo readings
%j1= -pi/2:pi/2         S1= 
%j2=pi/2:pi             S2=
%j_3=0:-pi/4
%j3=-pi/2: 13*pi/18     S3=

j1_home=0;
j1_min=-pi/2;
j1_max=pi/2;

j2_home=pi/2;
j2_min=11*pi/16;
j2_max=pi/8;

j3_home=-j2_home;
j3_min=-j2_min;
j3_max=-j2_max-(pi/3);
%j3=-j2+(j_3);

j4_home=-j2_home-j3_home;
j4_min=-j2_min-j3_min;
j4_max=-j2_max-j3_max;

% j4=-j2-j3;


%home
j_home=[j1_home j2_home j3_home ];
me_arm.plot(j_home)
pause(3)
hold on
%max
j_max=[j1_max j2_max j3_max];
me_arm.plot(j_max)
pause(3)

j_min=[j1_min j2_min j3_min];
me_arm.plot(j_min)
pause(3)

%%
for i=1:1000000
%you can generate N random numbers in the interval (a,b) with the formula r = a + (b-a).*rand(N,1).
%next step is to: define a range for each joint to move with, as this will
%be it max and min range of motion

theta_1=j1_min+(j1_max-j1_min)*rand();
theta_2=j2_min+(j2_max-j2_min)*rand();
theta_3=j3_min+(j3_max-j3_min)*rand();


m=me_arm.fkine([theta_1 theta_2 theta_3 ]);
%get the forward kinematics for each joint to get transformation matrix ,
% as we call it matrix m

me_arm.plot([theta_1 theta_2 theta_3]) %you can uncomment this

%from the matrix m we will extract the translation values as in matrix x
%(robot positions)
x(i,:)=(m.t)';
%the x variable stores the values x,y,z positions for a certain transfromation

s(i,:)=me_arm.ikine(m, [0 0 0], 'mask' ,[1 1 1 0 0 0]);
%make inverse kinematix for the output transformation to get and your joints
%values in matrix s 
%the s variable stores the joints values to make a certain rotaion


plot3(x(i,1), x(i,2), x(i,3),'o'); %you can uncomment this line and comment
% the next line

end
%%
%enter the positions

x=  0;
y=  22.8;
z= 3.15;
g=[1 0 0 x; 0 1 0 y; 0 0 1 z; 0 0 0 1];
me_i=me_arm.ikine(g, [0 0 0], 'mask' ,[1 1 1 0 0 0]);
me_arm.plot(me_i);
j1=me_i(1)
j2=me_i(2)
j3=me_i(3)

%%
%Servo map
min_base=120;
min_shoulder=70;
min_elbow=140;
min_gripper=50;
max_base=10;
max_shoulder=135;
max_elbow=100;
max_gripper=90;
home_base=60;
home_shoulder=90;
home_elbow=140;


servo_base=interp1([j1_min,j1_max],[min_base,max_base],me_i(1))
servo_elbow=interp1([j2_min,j2_max],[min_elbow,max_elbow],me_i(2))
servo_shoulder=interp1([j3_min,j3_max],[min_shoulder,max_shoulder],me_i(3))

%%
servo_publisher= rospublisher("/matlab_joints","geometry_msgs/Vector3")
joints_sent=rosmessage(servo_publisher);
joints_sent.X=servo_base;
joints_sent.Y=servo_shoulder
joints_sent.Z=servo_elbow
send(servo_publisher,joints_sent)

File: test_files/meArm.py
#this is the same as maltab code but in python
# generate a gui that take the x,y,z 
# and return the angles of the servos
# and the angles of the joints
# and the position of the end effector
# and the position of the base
# and the position of the shoulder
# and the position of the elbow
# and the position of the gripper
#use tkinter for the gui
#use matplotlib for the plot
# use spatialmath for the transformation
# use roboticstoolbox for the robot
# use numpy for the math
# use math for the math
# use matplotlib for the plot
# use tkinter for the gui


from roboticstoolbox import * 
import roboticstoolbox as rtb
from spatialmath.base import *
from spatialmath import *
import numpy as np
import math
import matplotlib.pyplot as plt
import tkinter

j1_max = np.pi/2
j2_max = np.pi/8
j3_max = -(np.pi/8)-(np.pi/3)

j1_min = -np.pi/2
j2_min = np.pi*(11/16)
j3_min = -np.pi*(11/16)

# mapping
min_base=120
min_shoulder=70
min_elbow=140
min_gripper=50
max_base=10
max_shoulder=135
max_elbow=100
max_gripper=90
home_base=60
home_shoulder=90
home_elbow=140


def map_range(x, old_min, old_max, new_min, new_max):
    old_range = old_max - old_min
    new_range = new_max - new_min
    return (((x - old_min) * new_range) / old_range) + new_min

class MeArm (DHRobot):
    def __init__(self):
        deg= np.pi/180
        L0 = RevoluteDH(d = 7.5, a=10, alpha=np.pi/2, qlim=[-np.pi/2, np.pi/2])
        L1 = RevoluteDH(d = 0, a=9, alpha=0, qlim=[np.pi*(11/16), np.pi/8])
        L2 = RevoluteDH(d = 0, a=9, alpha=0, qlim=[-np.pi*(11/16), -(np.pi/8)-(np.pi/3)])
        super().__init__([L0, L1, L2], name='MeArm' , manufacturer = "E-JUST")
        self.home = np.array([0, np.pi/2, -np.pi/2])
        self.qz = np.zeros(3)
        self.min = np.array([-np.pi/2, np.pi*(11/16), -np.pi*(11/16)])
        self.max = np.array([np.pi/2 , np.pi/8, -(np.pi/8)-(np.pi/3)])
        self.addconfiguration("home", self.home)
        self.addconfiguration("qz", self.qz)



if __name__=="__main__":
    meArm= MeArm()
    print(meArm)

t_home=meArm.fkine(meArm.home)    
print(t_home)

tz=meArm.fkine(meArm.qz)
print(tz.t)


qt = rtb.jtraj(meArm.qz, meArm.home, 50)

# meArm.plot(qt.q, backend='pyplot', movie='mearm.gif')
# meArm.plot( qt.q)

# meArm.plot(meArm.max, dt=1)
# x=meArm.fkine(meArm.home)
# print(x)
# meArm.plot(meArm.home, dt=40 )    
# meArm.q=meArm.home
# meArm.teach()

# my_ws = np.array([])

# for i in range (1,50,1):
#     theta_1 = j1_min + (j1_max - j1_min) * np.random.rand()
#     theta_2 = j2_min + (j2_max - j2_min) * np.random.rand()
#     theta_3 = j3_min + (j3_max - j3_min) * np.random.rand()
#     q=[theta_1, theta_2 , theta_3]
#     print(q)
#     meArm.plot(q)
#     fk = meArm.fkine(q)

# g = transl(0,22.8,3.15)


print("Enter X, Y, Z Positons: ")

x= float(input("X from 0 to 24: "))
print(type(x))
while x not in range(0,24):
    print("X is out of Workspace")
    x= float(input("X: "))

y=  float(input("Y from -24 to 24: "))
while y not in range(-24,24):
    print("Z is out of Workspace")
    y= float(input("Y: "))

z= int(input("Z from 2, 21: "))
while z not in range(2  ,21):
    print("Z is out of Workspace")
    z= float(input("Z: "))


g=SE3(x,y,z)
print(g)
s = meArm.ikine_LM(g)

global servo_base, servo_elbow, servo_shoulder
print("s.q",s.q)
if s.q[0]  > -np.pi/2 and s.q[0] < np.pi/2 :
    print("1")
    if s.q[1] > np.pi/8  and s.q[1] <  np.pi*(11/16):
        print('2')
        if s.q[2] > -np.pi*(11/16) and  s.q[2] < (-(np.pi/8)-(np.pi/3)):
            print('3')
            print(s.q)
            meArm.plot(s.q,dt=10)
            gk = meArm.fkine(s.q)
            print("gk",gk)

            servo_base = map_range(s.q[0], j1_min,j1_max,min_base, max_base)
            servo_shoulder = map_range(s.q[1], j2_min, j2_max, min_shoulder, max_shoulder)
            servo_elbow = map_range(s.q[2], j3_min, j3_max, min_elbow, max_elbow)

            print("Servo Base: ",servo_base,"Servo shoulder: ", servo_shoulder,"Servo Elbow: ", servo_elbow)



